File: frontend/src/mocks/fixtures/analysis.ts
import type { Analysis, AnalysisListItem } from "@/types/analysis.types";

export const mockAnalysis: Analysis = {
  id: 1,
  status: "completed",
  technique: { id: 1, display_name: "Jab", discipline: "Boxeo" },
  global_score: 73,
  power_score: 80,
  balance_score: 65,
  alignment_score: 78,
  speed_score: 70,
  xp_awarded: 30,
  joint_results: [
    {
      joint_name: "right_elbow",
      measured_angle: 145.2,
      reference_min: 165,
      reference_max: 180,
      optimal_angle: 175,
      is_correct: false,
      deviation: -29.8,
    },
    {
      joint_name: "right_shoulder",
      measured_angle: 88.0,
      reference_min: 80,
      reference_max: 100,
      optimal_angle: 90,
      is_correct: true,
      deviation: -2.0,
    },
  ],
  feedback: [
    {
      priority_order: 1,
      correction_title: "Extensión de codo insuficiente",
      correction_text: "Tu codo derecho alcanza 145°...",
      biomechanical_explanation: "La extensión completa del codo maximiza el alcance.",
      exercise_suggestion: "Practica shadow boxing con espejo.",
      impact_score: 0.85,
    },
  ],
  video_overlay_url: "/analysis/1/download/overlay",
  video_original_url: "/analysis/1/download/original",
  created_at: "2026-05-28T10:00:00",
  completed_at: "2026-05-28T10:01:30",
  error_message: null,
};

export const mockAnalysisListItem: AnalysisListItem = {
  id: 1,
  technique_display_name: "Jab",
  discipline_name: "Boxeo",
  global_score: 73,
  status: "completed",
  video_overlay_url: "/analysis/1/download/overlay",
  created_at: "2026-05-28T10:00:00",
};

export const mockAnalysisFailed: Analysis = {
  ...mockAnalysis,
  id: 2,
  status: "failed",
  global_score: null,
  xp_awarded: 0,
  error_message: "Video processing failed: no pose detected.",
};

export const mockAnalysisPending: Analysis = {
  ...mockAnalysis,
  id: 3,
  status: "pending",
  global_score: null,
  xp_awarded: 0,
  completed_at: null,
};

File: backend/app/schemas/analysis.py
"""
Module: schemas.analysis
Description: Pydantic schemas for analysis creation, detail responses, comparison and comments
"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


class TechniqueMinimal(BaseModel):
    id: int
    name: str
    display_name: str
    discipline_name: str
    difficulty: str

    model_config = {"from_attributes": True}


class JointResultSchema(BaseModel):
    joint_name: str
    measured_angle: float
    reference_min: float
    reference_max: float
    optimal_angle: float
    is_correct: bool
    deviation: float

    model_config = {"from_attributes": True}


class FeedbackItemSchema(BaseModel):
    priority_order: int
    correction_title: str
    correction_text: str
    biomechanical_explanation: Optional[str] = None
    exercise_suggestion: Optional[str] = None
    impact_score: float

    model_config = {"from_attributes": True}


class NewlyEarnedBadge(BaseModel):
    badge_id: int
    display_name: str
    xp_reward: int


class AnalysisDetailResponse(BaseModel):
    id: int
    status: str
    technique: Optional[TechniqueMinimal] = None
    global_score: Optional[float] = None
    power_score: Optional[float] = None
    balance_score: Optional[float] = None
    alignment_score: Optional[float] = None
    speed_score: Optional[float] = None
    xp_awarded: int = 0
    belt_upgraded: bool = False
    new_belt: Optional[str] = None
    newly_earned_badges: List[NewlyEarnedBadge] = []
    joint_results: List[JointResultSchema] = []
    feedback: List[FeedbackItemSchema] = []
    video_overlay_url: Optional[str] = None
    video_original_url: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime] = None
    error_message: Optional[str] = None

    model_config = {"from_attributes": True}


class AnalysisListItem(BaseModel):
    id: int
    technique_display_name: str
    discipline_name: str
    global_score: Optional[float] = None
    status: str
    video_overlay_url: Optional[str] = None
    created_at: datetime


class AnalysisListResponse(BaseModel):
    items: List[AnalysisListItem]
    total: int
    page: int
    limit: int
    pages: int


class CompareResponse(BaseModel):
    analysis_1: AnalysisDetailResponse
    analysis_2: AnalysisDetailResponse
    score_difference: float
    improved_joints: List[str]
    regressed_joints: List[str]
    improved: bool


class CommentCreate(BaseModel):
    content: str = Field(..., min_length=5, max_length=2000)


class CommentResponse(BaseModel):
    id: int
    content: str
    author_username: str
    author_avatar_url: Optional[str] = None
    created_at: datetime
